### DISCUSSION TITLE
Idea to improve library

### DISCUSSION THREAD

**Discussion by abuzuhri** [abuzuhri/Amazon-SP-API-CSharp#180] [Ideas]
Any one have idea for improve library are welcome write it here , I will be happy to have this library more and more used .
will do my best when I have free time

**Comment by J-W-Chan** [comment]
hi. [abuzuhri](https://github.com/abuzuhri)
i think we should find a better way to impletment multiple version for SP-API.
eg: 
1. CatalogItem API [v2020-12-01](https://developer-docs.amazon.com/sp-api/docs/catalog-items-api-v2020-12-01-reference#includeddata-subgroup-1) the parameter "includedData" is support the args "variations".
![image](https://user-images.githubusercontent.com/6493028/169196578-0adede8c-e286-4f92-80f2-531d861a7c29.png)
3. CatalogItem API [v2022-04-01](https://developer-docs.amazon.com/sp-api/docs/catalog-items-api-v2022-04-01-reference#includeddata-subgroup-1) the parameter "includedData" is not support the args "variations".
![image](https://user-images.githubusercontent.com/6493028/169196712-5ef801de-08be-409a-89d3-ce265c2481a2.png)

**Reply by J-W-Chan** [reply]
by the way. the api result is also particle different for the different version. @abuzuhri

**Reply by abuzuhri** [reply]
This weekend will work on this

**Comment by abuzuhri** [comment]
Here May Update and we need to be Updated 
https://github.com/amzn/selling-partner-api-docs/discussions/2289

**Reply by J-W-Chan** [reply]
I have completed the "getmyfeesestimates" api

**Comment by Paul-J-Barrett** [comment]
When someone originally took the swagger document to generate the library is it possible to now generate a swagger from the Library to compare to the original? I ask this because of the pull request I made to fix an issue in the FBAInbound GetShipments would probably show up in such a comparison.  Or has there just been to much deviation at this point?

I wish I knew swagger enough to do this but there are only so many hours in my days. It might be use full to have a swagger doc built during the build process then it could be compared by the community. Thoughts?

**Comment by J-W-Chan** [comment]
I have upgraded restsharp from 106 to 108, and synchronized changes from the main branch @abuzuhri 
[https://github.com/J-W-Chan/Amazon-SP-API-CSharp/tree/updateToRestsharp108](https://github.com/J-W-Chan/Amazon-SP-API-CSharp/tree/updateToRestsharp108)

**Reply by abuzuhri** [reply]
Please wait Don't push it I need to test it

**Reply by J-W-Chan** [reply]
> Please wait Don't push it I need to test it

Do we need new branches？

**Reply by abuzuhri** [reply]
I think it's better to have it in new branch

**Reply by J-W-Chan** [reply]
Ok. I've done a simple test. If the new branch is available, please help merge into the new branch

**Comment by nfzv** [comment]
### Would be great to get rid of Newtonsoft.Json
Thank you for the library, it is great, however the `Newtonsoft.Json` package that comes with it, may bring conflicts with `System.Text.Json.JsonSerializer`

This is of course a suggestion, but would be great to see.

**Newtonsoft** was replaced by Microsoft, and I believe its much preferable to use that, especially in packages.
[A little background](https://www.planetgeek.ch/2022/10/15/using-system-text-json-alongside-newtonsoft-json-net/)
[Might be useful](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/migrate-from-newtonsoft?pivots=dotnet-7-0)

It may require significant changes, haven't looked at the source yet, but will try my best to PR that sometime

**Comment by feetpeoplellc** [comment]
Suggestion:  Break reports out from the project.

Downloading reports could be its own project as it is isolated from other tasks and is basically the same for any report.  Maybe this would simplify the overall project for others, as well.  Simply because the process for reports is the same logic for all of them.

**Comment by brm121966** [comment]
Hi,
Pass the parameters of functions and procedures as an array or object, not as separate parameters.
Thanks.

**Comment by smargoli2** [comment]
Hi,
An option to add the User-Agent header on all requests would be nice. Can you point me in the direction of where I could possibly add this and make a PR?

**Comment by gilmore1996** [comment]
I'm not sure if I can't find it or if its not implemented but can you implement the /finances/2024-06-19/transactions GET method? This method is quite important now that deffered transactions aren't included in the shipment event list of the v0 Finance methods

**Comment by Paul-J-Barrett** [comment]
I have ran into this issue also and have tickets open with amazon. The two
events that are missing that we currently care about are: RemovalShipmentEvent
- Need for our various PBI dashboard on inventory.
TaxWithholdingEvent  - Need for our various tax reporting requirements
Unfortunately, I am not really interested in calling the API for individual
events. We currently ingest the main financesv0 feed into a datalake for
PBI reporting. Having to somehow call  individual events might work for
some but I don't think it would work for us with our volume.

**Comment by braddassow** [comment]
Ability to block an order from being shipped using Amazon Logistics (AMZL) in CreateFulfillmentOrderRequest. 
When reviewing Amazon Seller SP-API docs it states the FeatureSettings is part of the CreateFulfillmentOrderRequest. [https://developer-docs.amazon.com/sp-api/docs/fulfillment-outbound-api-v2020-07-01-reference#featuresettings]

`var featureSettings = new List<FeatureSettings>
        {
            new FeatureSettings
            {
                FeatureName = "BLOCK_AMZL",
                FeatureFulfillmentPolicy = FeatureFulfillmentPolicyEnum.Required,
            }
        };`

I have created featureSettings, but how can I attach it to the CreateFulfillmentOrderRequest?

**Comment by abuzuhri** [comment]
Hello @ProNotion I add you add as Collaborator in github to this project , please accept if you like to help on this project 

Any one else want to share the help here

**Comment by abuzuhri** [comment]
@kevinvenclovas i sent you invitation also

**Reply by kevinvenclovas** [reply]
Thank you @abuzuhri
